autoload/lsp.vim

```diff
@@ -148,28 +149,48 @@ function! s:unregister_events() abort
     doautocmd User lsp_unregister_server
 endfunction
 
+let s:event_queue = []
+let s:event_timer = -1
+
+function! s:add_event_queue(queue) abort
+    for l:queue in s:event_queue
+        if l:queue[0] == a:queue[0] && l:queue[1] == a:queue[1]
+            return
+        endif
+    endfor
+    call add(s:event_queue, a:queue)
+    call lsp#log('s:send_event_queue() will be triggered')
+    call timer_stop(s:event_timer)
+    let lazy = &updatetime > 1000 ? &updatetime : 1000
+    let s:event_timer = timer_start(lazy, function('s:send_event_queue'))
+endfunction
+
+function! s:send_event_queue(timer) abort
+    call lsp#log('s:send_event_queue()')
+    for l:queue in s:event_queue
+        for l:server_name in lsp#get_whitelisted_servers()
+            call s:ensure_flush(l:queue[0], l:server_name, l:queue[1])
+        endfor
+    endfor
+    let s:event_queue = []
+endfunction
+
 function! s:on_text_document_did_open() abort
     let l:buf = bufnr('%')
     call lsp#log('s:on_text_document_did_open()', l:buf, &filetype, getcwd(), lsp#utils#get_buffer_uri(l:buf))
-    for l:server_name in lsp#get_whitelisted_servers()
-        call s:ensure_flush(l:buf, l:server_name, function('s:Noop'))
-    endfor
+    call s:add_event_queue([l:buf, function('s:Noop')])
 endfunction
 
 function! s:on_text_document_did_save() abort
-    call lsp#log('s:on_text_document_did_save()', bufnr('%'))
     let l:buf = bufnr('%')
-    for l:server_name in lsp#get_whitelisted_servers()
-        call s:ensure_flush(bufnr('%'), l:server_name, {result->s:call_did_save(l:buf, l:server_name, result, function('s:Noop'))})
-    endfor
+    call lsp#log('s:on_text_document_did_save()', l:buf)
+    call s:add_event_queue([l:buf, {server_name, result->s:call_did_save(l:buf, server_name, result, function('s:Noop'))}])
```

Comment:
What happens if the event queue callback is called but the buffer is already removed?